### PR TITLE
feat: allow users to store stats as json to a file

### LINKED ISSUE #1535: Stats output without stdout
## Feature request
Generate a JSON stats file directly without forcing the user to redirect the stdout output

**What is the expected behavior?**
An option like --json that writes the output directly into a file

**What is motivation or use case for adding/changing the behavior?**
Build stats are a key feature for tracking (and linting) the quality of a build-pipeline. According the current documentation the only way to save the stats is [redirecting the stdout](https://webpack.js.org/api/stats/). 

The major problem is that when a user runs a manual build, it is not possible to get a build progress status because the progress line/text would mangle the json output.

**How should this be implemented in your opinion?**

1) add an option with the file output
--json-output [file]
--json-file [file]

2) Reference in the documentation the [Json Stats plugin]( https://www.npmjs.com/package/webpack-stats-plugin) 

Yes, there is already a plugin for this, but I believe the need for this feature justifies at a minimum a clear reference in the documentation about how to achieve this.

**Are you willing to work on this yourself?**
no

### PR DESCRIPTION
**Pull request by snitin315**
**What kind of change does this PR introduce?**
feature

**Did you add tests for your changes?**
Yes
**If relevant, did you update the documentation?**

**Summary**
Fixes #1535 

Screenshot at 2020-09-25 13-32-37

**Does this PR introduce a breaking change?**
No

**Other information**

### CODE REVIEW

File: packages/webpack-cli/lib/utils/Compiler.js [hunk 1]
```diff
@@ -57,8 +58,8 @@ class Compiler {
         if (!outputOptions.watch && (stats.hasErrors() || stats.hasWarnings())) {
             process.exitCode = 1;
         }
-        if (outputOptions.json) {
-            process.stdout.write(JSON.stringify(stats.toJson(outputOptions), null, 2) + '\n');
+        if (outputOptions.json === true) {
+            logger.raw(JSON.stringify(stats.toJson(outputOptions), null, 2));
```
Comment (evilebottnawi) [inline packages/webpack-cli/lib/utils/Compiler.js:62]: It should be `process.stdout.write`

Comment (evenstensberg) [reply packages/webpack-cli/lib/utils/Compiler.js:62]: +1

Comment (snitin315) [reply packages/webpack-cli/lib/utils/Compiler.js]: fixed

File: test/json/json.test.js [hunk 2]
```diff
@@ -13,6 +15,17 @@ describe('json flag', () => {
...
 
+    it('should store json to a file', (done) => {
+        const { stdout } = run(__dirname, ['--json', 'stats.json']);
+
+        expect(stdout).toContain('stats are succesfylly storerd as json to stats.json');
+        stat(resolve(__dirname, './stats.json'), (err, stats) => {
+            expect(err).toBe(null);
+            expect(stats.isFile()).toBe(true);
+            done();
+        });
```
Comment (evilebottnawi) [inline test/json/json.test.js:26]: expect(() => JSON.parse(file)).not.toThrow()

File: packages/webpack-cli/lib/utils/Compiler.js [hunk 3]
```diff
@@ -69,6 +70,15 @@ class Compiler {
                     statsErrors.push({ name: statErr.message, loc: errLoc });
                 });
             }
+            const JSONStats = JSON.stringify(stats.toJson(outputOptions), null, 2);
```
Comment (ylemkimon) [inline packages/webpack-cli/lib/utils/Compiler.js:73]: Note `JSON.stringify` doesn't end with a newline.

**Review by evenstensberg — DISMISSED** [review]

File: packages/webpack-cli/lib/utils/Compiler.js [hunk 4]
```diff
@@ -69,6 +70,15 @@ class Compiler {
...
+            if (typeof outputOptions.json === 'string') {
+                try {
+                    writeFileSync(outputOptions.json, JSONStats);
+                    logger.info(`stats are succesfylly storerd as json to ${outputOptions.json}`);
```
Comment (rishabh3112) [inline packages/webpack-cli/lib/utils/Compiler.js:77]:
```suggestion
                    logger.info(`stats are successfully stored as json to ${outputOptions.json}`);
```

File: packages/webpack-cli/lib/utils/cli-flags.js [hunk 5]
```diff
@@ -178,9 +178,9 @@ const core = [
     {
         name: 'json',
         usage: '--json',
-        type: Boolean,
+        type: [String, Boolean],
         alias: 'j',
-        description: 'Prints result as JSON',
+        description: 'Print result as JSON or store it in a file',
```
Comment (rishabh3112) [inline packages/webpack-cli/lib/utils/cli-flags.js:183]:
```suggestion
        description: 'Prints result as JSON or store it in a file',
```

File: test/json/json.test.js [hunk 6]
```diff
@@ -13,6 +15,22 @@ describe('json flag', () => {
...
+        readFile(resolve(__dirname, 'stats.json'), 'utf-8', (err, data) => {
+            expect(err).toBe(null);
+            expect(() => JSON.parse(data)).not.toThrow();
```
Comment (rishabh3112) [inline test/json/json.test.js:29]: Also add tests for few keys in parsed JSON.

**Review by rishabh3112 — CHANGES_REQUESTED** [review]
Left few suggestions.

File: packages/webpack-cli/README.md [hunk 7]
```diff
@@ -53,7 +53,7 @@ yarn add webpack-cli --dev
   --no-hot                 Disables Hot Module Replacement
   -d, --devtool string     Controls if and how source maps are generated.
   --prefetch string        Prefetch this request
-  -j, --json               Prints result as JSON
+  -j, --json string        Prints result as JSON or store it in a file
```
Comment (anshumanv) [inline packages/webpack-cli/README.md:56]: should be string, boolean right?

File: test/json/json.test.js (hunk 6 above)
Comment (rishabh3112) [reply test/json/json.test.js:33]: We should check these in object returned by `JSON.parse(data)`

File: packages/webpack-cli/README.md (hunk 7 above)
Comment (snitin315) [reply packages/webpack-cli/README.md]: hmm, updated

File: test/json/json.test.js (hunk 6 above)
Comment (snitin315) [reply test/json/json.test.js:35]: done

**Review by evilebottnawi — APPROVED** [review]

**Comment by evilebottnawi** [comment]
I think we can improve docs and replace this out of box https://github.com/FormidableLabs/webpack-stats-plugin

**Review by anshumanv — APPROVED** [review]

**Review by rishabh3112 — APPROVED** [review]

**Review by jamesgeorge007 — APPROVED** [review]

**Pull request merged**